[distributed] add processgroup options as argument

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#53663 [distributed] add processgroup options as argument**

This add the processgroup option as an optional argument to new_group
and init_processgroup, this allows user to pass in a initialized
processgroup option for gloo and nccl.

Also, to consolidate the APIs be more intuitive and easy to use, there're some Low-level API changes (which we expect the changes won't hit users):
* ProcessGroupGloo.Options.devices/threads ->
ProcessGroupGloo.Options._devices/_threads
* ProcessGroupNCCL.Options.op_timeout ->
ProcessGroupNCCL.Options.timeout
* ProcessGroupNCCL.Options.is_high_priority ->
ProcessGroupNCCL.Options.is_high_priority_stream
* Add ProcessGroupGloo.create_default_device
* Add constructors for ProcessGroupGloo/NCCL.Options

There might be some usages internally, which will be changed
together as part of this diff

fixes https://github.com/pytorch/pytorch/issues/50295

Differential Revision: [D26968857](https://our.internmc.facebook.com/intern/diff/D26968857)

wanchaol committed:
[distributed] add processgroup options as argument

This add the processgroup option as an optional argument to new_group
and init_processgroup, this allows user to pass in a initialized
processgroup option for gloo and nccl.

[ghstack-poisoned]

wanchaol committed:
Update on "[distributed] add processgroup options as argument"


This add the processgroup option as an optional argument to new_group
and init_processgroup, this allows user to pass in a initialized
processgroup option for gloo and nccl.

[ghstack-poisoned]

wanchaol committed:
Update on "[distributed] add processgroup options as argument"


This add the processgroup option as an optional argument to new_group
and init_processgroup, this allows user to pass in a initialized
processgroup option for gloo and nccl.

[ghstack-poisoned]

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -397,6 +398,7 @@ def init_process_group(backend,
                        world_size=-1,
                        rank=-1,
                        store=None,
+                       pg_options=None,
```
We should document this new argument, also, shouldn't all new args go at the end for BC?

wanchaol replied:
yeah should be, my fault, updating

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -559,6 +564,10 @@ def _new_process_group_helper(world_size,
         raise RuntimeError("Expected timeout argument to be of type"
                            "datetime.timedelta")
 
+    if pg_options is not None and timeout != default_pg_timeout:
```
Can we just check if they have the same timeout and continue? Otherwise, if different, raise an error indicating that they are conflicting, and note this in the doc?

wanchaol replied:
The check for this condition is actually quite complicated, the current check `if pg_options is not None and timeout != default_pg_timeout and pg_options.timeout != timeout:` mostly works, but if user pass in a `ProcessGroup.Options` with the timeout happen to be `options.timeout == default_pg_timeout`, this will trigger this runtime error, but it's still a legit input. I am wondering if we should just remove this check?

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -588,22 +597,36 @@ def _new_process_group_helper(world_size,
                 rank,
                 world_size,
-                timeout)
+                pg_options)
```
we should validate that this is an instance of `ProcessGroupNCCLOptions`, more generally should we validate that the options struct matches the backend and raise a useful error if it doesn't?

rohan-varma commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -1053,7 +1053,8 @@ that adds a prefix to each key inserted to the store.
           module, "ProcessGroupOptions")
-          .def_readwrite("backend", &::c10d::ProcessGroup::Options::backend);
+          .def_readwrite("backend", &::c10d::ProcessGroup::Options::backend)
+          .def_readwrite("timeout", &::c10d::ProcessGroup::Options::timeout);
```
We should document both the gloo options and nccl options including the arguments, defaults, and an example

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -588,22 +597,36 @@ def _new_process_group_helper(world_size,
+                pg_options.timeout = timeout
+                ifname_env = getenv("GLOO_SOCKET_IFNAME")
+                if ifname_env is not None:
+                    pg_options.devices = [ProcessGroupGloo.create_device(interface=iface) for iface in ifname_env.split(",")]
```
Does this mean if user creates their own `ProcessGroupGloo.Options()`, they are responsible for create device from interface name themselves? 

It looks like currently for gloo, we only provide `timeout` as a customizable option and always create devices in C++ similar to what we're doing here.

Should we remove `devices` arg from public interface of ProcessGroupGlooOptions and manually populate it always ourselves like we do currently? We can revisit allowing users to set it if there is a requirement.

rohan-varma reviewed (changes requested):
Looks great, thanks so much for taking this up! Had the following questions:

1) Are we only supporting nccl + gloo right now or want to support other backends such as mpi as well? Either way is probably fine but would be good to clarify

2) Can we add comprehensive unittests to the change? 

3) Can we add validation for ensuring the right options struct is passed in for the right backend?

4) Let's add documentation for all the new structs as well as new arg to `init_process_group`

5) For BC, should it be the last arg? 

6) More details inline, but could we remove devices arg from gloo options and always set those ourselves as we do currently?

wanchaol commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -1152,32 +1126,14 @@ that adds a prefix to each key inserted to the store.
-          &::c10d::ProcessGroupNCCL::Options::is_high_priority_stream)
-      .def_readwrite("op_timeout", &::c10d::ProcessGroupNCCL::Options::timeout);
+          "is_high_priority_stream",
+          &::c10d::ProcessGroupNCCL::Options::is_high_priority_stream);
```
yes correct.

yf225 commented:
Should we add a unit test for this change? Particularly very interested in how we can test the new NCCL process group is indeed using high priority stream now. :)

wanchaol committed:
Update on "[distributed] add processgroup options as argument"


This add the processgroup option as an optional argument to new_group
and init_processgroup, this allows user to pass in a initialized
processgroup option for gloo and nccl.

Differential Revision: [D26968857](https://our.internmc.facebook.com/intern/diff/D26968857)

[ghstack-poisoned]

wanchaol committed:
Update on "[distributed] add processgroup options as argument"


This add the processgroup option as an optional argument to new_group
and init_processgroup, this allows user to pass in a initialized
processgroup option for gloo and nccl.

Differential Revision: [D26968857](https://our.internmc.facebook.com/intern/diff/D26968857)

[ghstack-poisoned]

mrshenli reviewed (commented):
cc @pbelevich, are there any requirements/concerns from TLS support?

pritamdamania87 commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -1191,12 +1224,41 @@ that adds a prefix to each key inserted to the store.
+      )")
+      .def(
+          py::init<std::chrono::milliseconds, bool>(),
+          py::arg("timeout") = kProcessGroupDefaultTimeout,
```
Why do we have a `timeout` arg in ProcessGroupOptions if both `init_process_group` and `new_group` already have a timeout?

wanchaol replied:
I think mostly from an intuition perspective, a timeout is actually an option to construct the process group, and I think it might make some sense to group all the options for ProcessGroup together in a single struct. ProcessGroupNCCL and ProcessGroupGloo both have it, so I think we might just put this into the base class and expose to the user. Do you mean that we better find a way to hide the `timeout` arg from the user and set it internally?

rohan-varma replied:
I think we should have it in the options as it's an option to build the process group, but agreed that we should figure out the precedence/how to handle when both values are set.

pritamdamania87 replied:
Sorry about the delay here, but I don't think we should have two timeouts in two different places in the API. This is confusing for users. Is the goal of this feature to retrieve the options from an existing pg and use those in `new_group` to create a new group with the same options? If so, shouldn't `pg_options` be an argument only in `new_group` and not in `init_process_group`?

Can we maybe assert that `timeout` and `pg_options` should not be specified together and the user should use only one of them?

I guess the same applies to `backend` as well? That option is now present in the API and in pg_options as well. 

cc @rohan-varma @wanchaol

wanchaol replied:
Yeah agree that we should have a unified place to put in timeout. The goal of this feature is to enable the options to be available during the creation of process group, I think both `init_process_group` and `new_group` should have this supported? since some user might only want one processgroup but configure it with the options during initialization (i.e. user only creates a nccl processgroup with high_priority_stream). Given that we don't want to confuse user for setting different timeouts, the current approach is that if we detect that user pass in timeouts and specify timeouts in ProcessGroupOptions, we error out to the user about this behavior about timeout conflicting. 

I think assert timeout and pg_options be not specified together sounds good to me, it's actually a much more clean approach. The only question is timeout is not an optional argument as of now (it uses default value), so we couldn't know if two are both specified or not (we can check if timeout uses default_pg_timeout though but that will not totally avoid it).  Let me try changing timeout to `Optional[timedelta]` in both `init_process_group` and `new_group`, I think it should be fine as we can specify default_pg_timeout inside the APIs (should not be BC breaking).

wanchaol commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -1191,12 +1224,41 @@ that adds a prefix to each key inserted to the store.
+          py::arg("timeout") = kProcessGroupDefaultTimeout,
+          py::arg("is_high_priority_stream") = false)
       .def_readwrite(
-          "is_high_priority",
```
do people think we should change this API to `is_high_priority_stream` instead of `is_high_priority`? this might cause some internal code change, but given that OSS user didn't aware this api yet (we didn't put it in the doc yet). I think it's relatively safe to be not BC breaking?

rohan-varma replied:
I think naming it `is_high_priority_stream` should be good, since we don't expose it there should be no BC concerns

wanchaol committed:
Update on "[distributed] add processgroup options as argument"


This add the processgroup option as an optional argument to new_group
and init_processgroup, this allows user to pass in a initialized
processgroup option for gloo and nccl.

Differential Revision: [D26968857](https://our.internmc.facebook.com/intern/diff/D26968857)

[ghstack-poisoned]

rohan-varma commented on test/distributed/test_c10d.py:
```diff
@@ -4639,6 +4640,48 @@ def test_broadcast_coalesced_gloo_cpu(self):
+    @requires_gloo()
+    def test_pass_gloo_options(self):
+        pg_opts = c10d.ProcessGroupGloo.Options()
+        pg_opts.devices = [create_device(interface=LOOPBACK)]
```
Can we somehow test that we appropriately set devices if the user does not set it themselves?

wanchaol replied:
Yeah that's also something I would like to test, but I don't actually know how to test the devices/threads attribute creation yet, given that we don't hold a `options` field in ProcessGroupGloo and there's currently no way for us to access the protected fields in python (from the [python bindings](https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/init.cpp#L1125) currently, we don't have access to fields like `context_` and `threads_`). So unless we want to expose them in python or add a new field in ProcessGroupGloo and bind it in python, we couldn't know if we set up the devices or threads?

rohan-varma commented on test/distributed/test_c10d.py:
```diff
@@ -4639,6 +4640,48 @@ def test_broadcast_coalesced_gloo_cpu(self):
+            pg_options=pg_opts
+        )
+
+        default_pg = c10d.distributed_c10d._get_default_group()
```
Can we verify that all getters for the pg options work appropriately?

rohan-varma commented on test/distributed/test_c10d.py:
```diff
@@ -4639,6 +4640,48 @@ def test_broadcast_coalesced_gloo_cpu(self):
+            pg_options=pg
+        )
+
+        default_pg = c10d.distributed_c10d._get_default_group()
```
nit: unused?

rohan-varma commented on test/distributed/test_c10d.py:
```diff
@@ -4639,6 +4640,48 @@ def test_broadcast_coalesced_gloo_cpu(self):
+        t = torch.tensor([self.rank + 1] * 10).cuda(2 * self.rank)
+        pg.allreduce(t).wait()
+        expected_tensor = torch.tensor([3] * 10).cuda(2 * self.rank)
+        self.assertEqual(expected_tensor, t)
```
Does this just verify allreduce is correct or does it also somehow verify we're using a high-pri stream?

rohan-varma commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -1098,27 +1104,54 @@ that adds a prefix to each key inserted to the store.
+    >>> gloo_options = dist.ProcessGroupGloo.Options(timeout=timedelta(seconds=300))
+    >>> # initialize a gloo process group with the options just created
+    >>> dist.init_process_group("gloo", init_method=f"file://{temp_name}",
+    >>>                          world_size=2, rank=0, pg_options=gloo_options)
```
Thanks for adding the docs!

rohan-varma replied:
Can we test the functionality when timeout is set in options and provided as an arg?

rohan-varma commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -1158,7 +1191,7 @@ that adds a prefix to each key inserted to the store.
           py::arg("rank"),
           py::arg("size"),
-          py::arg("timeout") = std::chrono::milliseconds(10 * 1000), // NOLINT
+          py::arg("timeout") = kProcessGroupDefaultTimeout, // NOLINT
```
nit: might be able to get rid of NOLINT if it was due to magic number

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -451,6 +453,12 @@ def init_process_group(backend,
+        pg_options (ProcessGroupOptions, optional): process group options
+            specifying what additional options need to be passed in during
+            the construction of specific process groups. i.e. for the ``nccl``
+            backend, is_high_priority_stream can be specified so that process
```
nit: backtick high_priority_stream

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -588,22 +604,48 @@ def _new_process_group_helper(world_size,
+                else:
+                    pg_options.devices = [ProcessGroupGloo.create_default_device()]
+
+                pg_options.threads = len(pg_options.devices) * 2
```
So looks like `threads` is only set if not pg_options.devices, what happens if user sets `devices` but not threads? Should we throw error or set it for them?

rohan-varma replied:
Looks like it's not too much of an issue as we expose neither threads nor device in the docs. Although, we should probably do something reasonable in any permutation of args we're given, or maybe prefix `threads` and `devices` with `_`?

wanchaol replied:
I think making them be prefixed with `_` is good as this is something we don't want to expose to the user yet, and when we construct them or user choose the construct them, they should ensure both options are set. Probably need to change the internal code to use prefix as well. 

For permutation of args, I am thinking of the following pattern:
* pg_options._device=None, pg_options._threads=default -> we generate these
* pg_options._device=None, pg_options._threads!= default -> user pass in a option with threads set but not options, error to the user
* pg_options._device!=None, pg_options._threads!=default -> legit options passed by user
* pg_options._device!=None, pg_options._threads=default -> we treat this as legit as well and it's user's responsibility to ensure it's correct.

rohan-varma replied:
Awesome! This sounds good to me.

rohan-varma commented on test/distributed/test_c10d.py:
```diff
@@ -4639,6 +4640,48 @@ def test_broadcast_coalesced_gloo_cpu(self):
+        pg_opts = c10d.ProcessGroupGloo.Options()
+        pg_opts.devices = [create_device(interface=LOOPBACK)]
+        pg_opts.timeout = timedelta(seconds=1)
+        pg_opts.threads = 2
```
Maybe we want to test what happens if devices is set but threads are not, and if threads are set but devices or not?  Although doesn't seem like too big of an issue - we are only exposing timeout so this logic should be okay.

rohan-varma reviewed (commented):
Looks great! Just had some more questions mostly around devices and threads for gloo. Otherwise I think it should be good to go.

rohan-varma commented:
cc @srinivas212 - who I believe was also requesting the feature offline to see if this has all requirements

wanchaol committed:
Update on "[distributed] add processgroup options as argument"


This add the processgroup option as an optional argument to new_group
and init_processgroup, this allows user to pass in a initialized
processgroup option for gloo and nccl.

Differential Revision: [D26968857](https://our.internmc.facebook.com/intern/diff/D26968857)

[ghstack-poisoned]

wanchaol committed:
Update on "[distributed] add processgroup options as argument"


This add the processgroup option as an optional argument to new_group
and init_processgroup, this allows user to pass in a initialized
processgroup option for gloo and nccl.

Also, to consolidate the APIs be more intuitive and easy to use, there're some Low-level API changes (which we expect the changes won't hit users):
* ProcessGroupGloo.Options.devices/threads ->
ProcessGroupGloo.Options._devices/_threads
* ProcessGroupNCCL.Options.op_timeout ->
ProcessGroupNCCL.Options.timeout
* ProcessGroupNCCL.Options.is_high_priority ->
ProcessGroupNCCL.Options.is_high_priority_stream
* Add ProcessGroupGloo.create_default_device
* Add constructors for ProcessGroupGloo/NCCL.Options

There might be some usages internally, which will be changed
together as part of this diff

Differential Revision: [D26968857](https://our.internmc.facebook.com/intern/diff/D26968857)

[ghstack-poisoned]

wanchaol committed:
add options and timeout test case on "[distributed] add processgroup options as argument"


This add the processgroup option as an optional argument to new_group
and init_processgroup, this allows user to pass in a initialized
processgroup option for gloo and nccl.

Also, to consolidate the APIs be more intuitive and easy to use, there're some Low-level API changes (which we expect the changes won't hit users):
* ProcessGroupGloo.Options.devices/threads ->
ProcessGroupGloo.Options._devices/_threads
* ProcessGroupNCCL.Options.op_timeout ->
ProcessGroupNCCL.Options.timeout
* ProcessGroupNCCL.Options.is_high_priority ->
ProcessGroupNCCL.Options.is_high_priority_stream
* Add ProcessGroupGloo.create_default_device
* Add constructors for ProcessGroupGloo/NCCL.Options

There might be some usages internally, which will be changed
together as part of this diff

Differential Revision: [D26968857](https://our.internmc.facebook.com/intern/diff/D26968857)

[ghstack-poisoned]

wanchaol committed:
fix windows test on "[distributed] add processgroup options as argument"


This add the processgroup option as an optional argument to new_group
and init_processgroup, this allows user to pass in a initialized
processgroup option for gloo and nccl.

Also, to consolidate the APIs be more intuitive and easy to use, there're some Low-level API changes (which we expect the changes won't hit users):
* ProcessGroupGloo.Options.devices/threads ->
ProcessGroupGloo.Options._devices/_threads
* ProcessGroupNCCL.Options.op_timeout ->
ProcessGroupNCCL.Options.timeout
* ProcessGroupNCCL.Options.is_high_priority ->
ProcessGroupNCCL.Options.is_high_priority_stream
* Add ProcessGroupGloo.create_default_device
* Add constructors for ProcessGroupGloo/NCCL.Options

There might be some usages internally, which will be changed
together as part of this diff

Differential Revision: [D26968857](https://our.internmc.facebook.com/intern/diff/D26968857)

[ghstack-poisoned]

wanchaol committed:
Update on "[distributed] add processgroup options as argument"


This add the processgroup option as an optional argument to new_group
and init_processgroup, this allows user to pass in a initialized
processgroup option for gloo and nccl.

Also, to consolidate the APIs be more intuitive and easy to use, there're some Low-level API changes (which we expect the changes won't hit users):
* ProcessGroupGloo.Options.devices/threads ->
ProcessGroupGloo.Options._devices/_threads
* ProcessGroupNCCL.Options.op_timeout ->
ProcessGroupNCCL.Options.timeout
* ProcessGroupNCCL.Options.is_high_priority ->
ProcessGroupNCCL.Options.is_high_priority_stream
* Add ProcessGroupGloo.create_default_device
* Add constructors for ProcessGroupGloo/NCCL.Options

There might be some usages internally, which will be changed
together as part of this diff

Differential Revision: [D26968857](https://our.internmc.facebook.com/intern/diff/D26968857)

[ghstack-poisoned]

rohan-varma commented on test/distributed/test_c10d.py:
```diff
@@ -4690,6 +4691,88 @@ def test_broadcast_coalesced_gloo_cpu(self):
+
+        default_pg = c10d.distributed_c10d._get_default_group()
+
+        # Test properly set devices on options if user don't set devices
```
Nice, thanks for adding these tests!

rohan-varma commented on test/distributed/test_c10d.py:
```diff
@@ -4690,6 +4691,88 @@ def test_broadcast_coalesced_gloo_cpu(self):
+        # Test with new_group
+        pg = c10d.new_group([0, 1], pg_options=pg_opts)
+        # test if the process group constructed with high priority stream
+        self.assertEqual(pg.options.is_high_priority_stream, True)
```
nit: just self.assertTrue

rohan-varma commented on test/distributed/test_c10d.py:
```diff
@@ -4690,6 +4691,88 @@ def test_broadcast_coalesced_gloo_cpu(self):
+            world_size=self.world_size,
+            rank=self.rank,
+            store=store,
+            pg_options=pg
```
Confused as to how this passes, looks like `pg` is undefined here and it should be pg_opts?

wanchaol replied:
this seems weird.. looking into what's going on there.

wanchaol replied:
something weird with `@skip_if_not_multigpu` and made the test case not run as expected.. change it to `    @skip_if_lt_x_gpu(2)` and verified its working.

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -492,6 +500,7 @@ def init_process_group(backend,
             [],
             Backend.MPI,
             None,
+            pg_options=pg_options,
```
This is creating MPI process group right, which doesn't support options? Is this codepath exercised by tests anywhere?

wanchaol replied:
Yeah make sense, probably should remove it. I was just passing in `pg_options` to `_new_process_group_helper`, and `_new_process_group_helper` only use `pg_options` for ProcessGroupGloo and NCCL, so MPI pg will just ignore pg_options.

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -559,6 +570,11 @@ def _new_process_group_helper(world_size,
         raise RuntimeError("Expected timeout argument to be of type"
                            "datetime.timedelta")
 
+    if pg_options is not None and timeout != default_pg_timeout and pg_options.timeout != timeout:
```
so if pg_options has a non-default timeout set, but user did not set timeout and thus timeout == default_pg_timeout, it looks like we override it with the pg_options timeout? This is probably fine, but would be good to test and document this.

wanchaol replied:
yes, sounds good let me add this behavior to the doc.

rohan-varma reviewed (approved):
Looks good! Just some misc questions/suggestions, then it should be good to go.

wanchaol committed:
Update on "[distributed] add processgroup options as argument"


This add the processgroup option as an optional argument to new_group
and init_processgroup, this allows user to pass in a initialized
processgroup option for gloo and nccl.

Also, to consolidate the APIs be more intuitive and easy to use, there're some Low-level API changes (which we expect the changes won't hit users):
* ProcessGroupGloo.Options.devices/threads ->
ProcessGroupGloo.Options._devices/_threads
* ProcessGroupNCCL.Options.op_timeout ->
ProcessGroupNCCL.Options.timeout
* ProcessGroupNCCL.Options.is_high_priority ->
ProcessGroupNCCL.Options.is_high_priority_stream
* Add ProcessGroupGloo.create_default_device
* Add constructors for ProcessGroupGloo/NCCL.Options

There might be some usages internally, which will be changed
together as part of this diff

Differential Revision: [D26968857](https://our.internmc.facebook.com/intern/diff/D26968857)

[ghstack-poisoned]

wanchaol committed:
Update on "[distributed] add processgroup options as argument"


This add the processgroup option as an optional argument to new_group
and init_processgroup, this allows user to pass in a initialized
processgroup option for gloo and nccl.

Also, to consolidate the APIs be more intuitive and easy to use, there're some Low-level API changes (which we expect the changes won't hit users):
* ProcessGroupGloo.Options.devices/threads ->
ProcessGroupGloo.Options._devices/_threads
* ProcessGroupNCCL.Options.op_timeout ->
ProcessGroupNCCL.Options.timeout
* ProcessGroupNCCL.Options.is_high_priority ->
ProcessGroupNCCL.Options.is_high_priority_stream
* Add ProcessGroupGloo.create_default_device
* Add constructors for ProcessGroupGloo/NCCL.Options

There might be some usages internally, which will be changed
together as part of this diff

Differential Revision: [D26968857](https://our.internmc.facebook.com/intern/diff/D26968857)

[ghstack-poisoned]
